p2panda-rs/src/document/document_view.rs

```diff
@@ -1,251 +1,206 @@
-// SPDX-License-Identifier: AGPL-3.0-or-later
+// SPDX-License-Identifier: AGPL-3.view-or-later
 
 //! Types and methods for deriving and maintaining materialised documents.
 use std::collections::btree_map::Iter as BTreeMapIter;
 use std::collections::BTreeMap;
-use std::convert::TryFrom;
 
-use crate::document::error::DocumentViewError;
-use crate::operation::{AsOperation, Operation, OperationValue, OperationWithMeta};
+use crate::hash::Hash;
+use crate::operation::OperationValue;
+
+/// The ID of a document view. Contains the hash id of the document, and the hash ids of the
+/// current document graph tips.
+#[derive(Debug, PartialEq, Clone)]
+pub struct DocumentViewId {
+    document_id: Hash,
+    view_id: Vec<Hash>,
+}
+
+impl DocumentViewId {
+    /// Create a new document view id.
+    pub fn new(document_id: Hash, view_id: Vec<Hash>) -> Self {
+        Self {
+            document_id,
+            view_id,
+        }
+    }
+
+    /// Get just the document id.
+    pub fn document_id(&self) -> &Hash {
+        &self.document_id
+    }
+
+    /// Get just the view id.
+    pub fn view_id(&self) -> &[Hash] {
+        self.view_id.as_slice()
+    }
+}
 
 /// The materialised view of a `Document`. It's fields match the documents schema definition.
 ///
 /// `DocumentViews` can be instantiated from a CREATE operation and then mutated with UPDATE
 /// or DELETE operations.
-#[derive(Debug, PartialEq, Default, Clone)]
-pub struct DocumentView(BTreeMap<String, OperationValue>);
+#[derive(Debug, PartialEq, Clone)]
+pub struct DocumentView {
+    pub(crate) id: DocumentViewId,
+    pub(crate) view: BTreeMap<String, OperationValue>,
```

Comment:
Can `String` here also be expressed as `FieldType`? (And in similar places)

Reply:
oooh, yes, I think we talked about that already.

Reply:
Interesting naming here: Is every `BTreeMap<FieldType, OperationValue>` a _view_?

Reply:
You mean the `FieldType` naming? That should actually be `FieldName` or similar.

Or did you mean something else?

Reply:
Still having `OperationValue` is strange too, but that's what it is....

Reply:
> Still having `OperationValue` is strange too, but that's what it is....

One could say that with materialization the "last operation value wins"? 🤔😅

Reply:
> You mean the `FieldType` naming? That should actually be `FieldName` or similar.
> 
> Or did you mean something else?

Oh, right.